### DISCUSSION TITLE
Export u3d in a container?

### DISCUSSION THREAD

**Discussion by ghost** [cnr-isti-vclab/PyMeshLab#25] [Q&A] [answered]
Hello PyMeshLab Team!

What do you think it would take to be able to use pymeshlab inside a container spawned from the latest ubuntu image (or something like that). Is this possible? I am currently attempting to do it with xvfb but would appreciate any suggestions.

Thanks!

**Comment by alemuntoni** [comment]
Hi @bdf-imbio! 
I am not really into containers, but if python and numpy can be installed, then it should work.
As far as I know, github actions works with containers, and all the tests are run on it!
Are you experiencing issues?

**Reply by ghost** [reply]
I am having trouble getting things to work so I will explain my goal and maybe you can point me in the right direction:

**Goal:** Convert a `.ply` file to a `.u3d` file for use in a `pdf`.

I was able to do that easily on my macbookpro using the meshlab gui program but now I need to do it in a container as part of a larger project.

Googling, I tried `meshlabserver` (never actually got it to work) and eventually found `pymeshlab`. 

The container is `ubuntu:18.04` from dockerhub and to that I added `git` (with `apt`) and `python3.7` and `pip` (with `conda`).  

Then I ran the following from the command prompt within the running container:
`pip install git+https://github.com/cnr-isti-vclab/PyMeshLab`. That completes without errors.

Next, I launch `python` and try to import `pymeshlab`. When I do so, I get the following error:
```python
>>> import pymeshlab as ml
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/miniconda/lib/python3.7/site-packages/pymeshlab/__init__.py", line 1, in <module>
    from .pmeshlab import *
ImportError: libGL.so.1: cannot open shared object file: No such file or directory
```

Based on the message, I am missing a graphics library. Instead of trying to install that specific library, I figured the `meshlab` package for ubuntu would have it and installed it with `apt install meshlab`. When done, I can import `pymeshlab`:

```python
>>> import pymeshlab as ml
PyMeshLab 0.1.6 based on MeshLab 2020.09d
```
(I did notice that ubuntu installs 2020.09d instead of your latest, maybe that's an issue?)

From there, I follow your example in the readme with the following output:
(Note, `simple.ply` is an ascii file made with Meshlab on my mbp)
```python
Python 3.7.8 | packaged by conda-forge | (default, Nov 27 2020, 19:24:58)
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pymeshlab as ml
PyMeshLab 0.1.6 based on MeshLab 2020.09d
>>> ms = ml.MeshSet()
>>> ms.load_new_mesh('simple.ply')
>>> ms.save_current_mesh('simple.u3d')
Must construct a QGuiApplication first.
Exit code = 80000000
QWidget: Must construct a QApplication before a QWidget
Aborted
```
No output is created and the interactive `python` session is killed.

Any idea what's happening there? Based on the message, it seems to be requiring a gui application, am I understanding that correctly?

Thanks for any help you can give :)

**Reply by ghost** [reply]
As an experiment, I tried doing the conversion above on my mbp as it has a working gui and meshlab 2020.12. The result was similar:

```python
Python 3.7.9 (default, Nov 18 2020, 13:03:57)
[Clang 12.0.0 (clang-1200.0.32.27)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import pymeshlab as ml
PyMeshLab 0.1.6 based on MeshLab 2020.09d
>>> ms = ml.MeshSet()
>>> ms.load_new_mesh('inputs/simple.ply')
>>> ms.save_current_mesh('yolo.u3d')
Must construct a QGuiApplication first.
Exit code = 80000003
QWidget: Must construct a QApplication before a QWidget
Abort trap: 6
```

**Reply by alemuntoni** [reply]
You are doing everything right!
Here there are two main issues:
- dependency of libGL: this is an old issue also with meshlabserver. It is a matter of big reorganization of code in the "meshlab side" - separate everything that needs graphic libraries from everything that don't. It is a big TODO in our list :) right now the solution is just to have installed a graphic library!
- u3d: the u3d exporter it is very difficult to make it working. I actually did not test it in 0.1.6 (my bad), therefore right now just does not work (on mac and linux, it should on windows but I need to check). 

In any case, thank you for reporting this! I'll try to solve at least the u3d exporter for the next pymeshlab version!

**Reply by ghost** [reply]
Hey, thanks for getting back to me.

If I read you right, conversion to the `u3d` format isn't possible with `pymeshlab` for the time being?

**Reply by alemuntoni** [reply]
exactly, right now it does not work. I'll try to fix it asap, but I'm afraid that it will take some time.

**Reply by ghost** [reply]
I totally understand, thanks for your time and thanks for (py)meshlab!

**Reply by ghost** [reply]
I would still like to convert to `u3d`, do you have any suggestions on how to go about that? How does meshlab do it? 

The tutorials I've found (all at least a few years old) point to [u3d](https://sourceforge.net/projects/mathgl/files/U3D%20tools/) but I find that even when I get stuff to seemingly build correctly, it doesn't convert.

**Reply by ghost** [reply]
I did some looking today on the `meshlab` github and eventually found [this](https://github.com/ningfei/u3d) which I believe is the converter used by meshlab. With it, I tried the following in my container:

```bash
$ IDTFConverter -dl 1 dbg.txt -i simple.idtf -o simple.u3d
Exit code = 80000000
```

As you can see, the error code isn't very helpful. I tried building `u3d` from source and installing with conda. I tried with `idtf` produced by `meshlab` and packaged with `u3d`. Same result every time.

I know you aren't responsible for that software but maybe you have some idea of what's going on? I am starting to think its broken.

Thanks

**Reply by alemuntoni** [reply]
Yes, that repo you linked is the same that meshlab uses, with some modifications.
This u3d exporter it's a huge mess. Unlike any other reasonable library, the IFX libraries are not loaded using the linker, but are loaded internally in the C++ code of the converter. This makes impossible to change anything in the library, and it is very easy to break something.
The only way to understand what's going on is to debug the code and look which library is failing to be loaded (and try to understand why).

I know that the u3d exporter is a very appealing feature of meshlab, but managing this library is very painful, and unfortunately there are no alternatives.

**Reply by alemuntoni** [reply]
Moreover, loading a library internally in the C++ code, means that the starting directory where the search of the library is performed, is the same of the binary that is being executed. That binary, in the pymeshlab case, is the call of python... Which may be literally anywhere in the system. That's why in pymeshlab the exporter is not working.
I am sorry that the u3d exporter does not work, but fixing this would take a lot of time that I just don't have right now. I'll fix this whenever it will be possible, but I can't make any promise at the moment :(

**Reply by ghost** [reply]
Thanks for explaining all that. I understand you are super busy so I really appreciate the time you've taken to walk me through this stuff. Whenever you do get around to making the u3d export work, I have a project for it :)

**Reply by Sea421** [reply]
Hi @bdf-imbio,

I have faced with same issue. Please visit the [discussion](https://github.com/cnr-isti-vclab/meshlab/discussions/871#discussioncomment-234151).
In my case it was locale issue. May it help

**Reply by alemuntoni** [reply]
u3d exporter now works fine in PyMeshLab! :)
The fix will be available in the next releases!

**Comment by ghost** [comment, marked as answer]
It is possible to export u3d files with pymeshlab in a container but it requires installing the IDTFConverter tool manually.

Here is a Dockerfile with the requisite dependencies and steps:
```Docker
FROM ubuntu

# install pymeshlab dependencies
# noninteractive frontend required to install tzdata w/out input
RUN apt update && DEBIAN_FRONTEND=noninteractive apt install --yes \
    python3 \
    python3-pip \
    libgl1-mesa-dev \
    tzdata \
    libglib2.0-0
# install pymeshlab
RUN pip3 install pymeshlab

# install IDTFConverter
# dependencies
RUN apt update && apt install --yes \
    build-essential \
    git \
    libpng-dev \
    libjpeg-dev \
    locales
# set local to en_US.UTF-8 (IDTFConverter expects this)
RUN sed -i '/en_US.UTF-8/s/^# //g' /etc/locale.gen && locale-gen
# clone u3d repo
RUN git clone https://github.com/ningfei/u3d.git /tmp/u3d
# build
WORKDIR /tmp/u3d
RUN ./configure && make && make install

WORKDIR /root/input
```
Build an image from the Dockerfile with `docker build -t u3d_exporter .` from the directory containing the Dockerfile.

Spawn a container from said image with `docker run -it --rm -v /path/to/your/meshes:/root/input u3d_exporter` where the `-v` option allows you to share your local meshes with the container.

From the container prompt, launch `python3`, import `pymeshlab` and export to `u3d`:
```python
import pymeshlab as ml
ms = ml.MeshSet()
ms.load_new_mesh('your.mesh')
...
ms.save_current_mesh('your_mesh.u3d')
```

You should now have a `u3d` file you can use to with a `3Dpdf`. For questions concerning a specific step or tool, please consult the relevant documentation.

Thanks to @Sea421 for figuring out the `locale` issue that was breaking the `IDTFConverter`.

Note that at this time, color `u3d` output is not possible, likely related to #42.

Also, at some point the steps above related to installing the `IDTFConverter` maybe become irrelevant when `pymeshlab` is fixed to do the conversion itself, see #27.
